### PR TITLE
Refactor prometheusreceiver config to avoid some custom unmarshal work

### PR DESCRIPTION
**Pull request by bogdandrutu**
This is **NOT** a user config breaking change.

This PR improves the usability of our public APIs removing unnecessary placeholders public members. For API developers they only need to interact with the new wrapper types and don't worry about placeholders.

Also this PR simplifies the unmarshalling logic allowing to add more fields more confidently.

### CODE REVIEW

**Review by dashpole — APPROVED** [review]

**Review by Aneurysm9 — CHANGES_REQUESTED** [review]
If this is a breaking change it needs significantly more description of its purpose and effect.  What are we trying to do here? Why is it necessary and advisable? What impact does it have on users (both end users and developers of components that interact with this component)?  There's no description on this PR nor is there an issue detailing what needs to change and why.

**Comment by bogdandrutu** [comment]
@Aneurysm9 tried to clarify what happens. It is a clear improvement in my opinion since we remove the hacky placeholders.

**Review by Aneurysm9 — DISMISSED** [review]
Still reviewing this but I'm on-call today and Monday and unlikely to be able to finish a review before Tuesday.

**Comment by Aneurysm9** [comment]
> It is an API breaking change because it changes a public type. It isn't a breaking change for users.

API users are still users.  Has this been communicated to @open-telemetry/operator-ta-maintainers?  I know they're at least one group that is going to have an interest in this.

File: receiver/prometheusreceiver/config.go
Comment (bogdandrutu) [reply receiver/prometheusreceiver/config.go:60]: Good call.

File: receiver/prometheusreceiver/config.go
Comment (bogdandrutu) [reply receiver/prometheusreceiver/config.go:48]: Not sure I agree with your change, because that will be a behavior change (which I am avoiding in a refactor PR), here is the previous code:

```golang
func (cfg *Config) Validate() error {
	promConfig := cfg.PrometheusConfig
	if promConfig != nil {
		err := cfg.validatePromConfig(promConfig)
		if err != nil {
			return err
		}
	}
	...
}

func (cfg *Config) validatePromConfig(promConfig *promconfig.Config) error {
	if len(promConfig.ScrapeConfigs) == 0 && cfg.TargetAllocator == nil {
		return errors.New("no Prometheus scrape_configs or target_allocator set")
	}
	....
}
```

You can see that if the  `promConfig` is `nil` will not be an error in the previous code. I understand that my refactored code makes it more obvious that this may be a bug, but we can have a fix in a followup PR.

**Comment by bogdandrutu** [comment]
> API users are still users. 

As per my understanding in the changelog we use `api` and `user` to differentiate between API users and end-user configuration/behavior breaking changes. If you disagree with these usages of the 2 words we can change that.

> Has this been communicated to @open-telemetry/operator-ta-maintainers? I know they're at least one group that is going to have an interest in this.

I don't see a need to do that, since https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-operator%20prometheusreceiver&type=code doesn't show any usage of the public API.

File: receiver/prometheusreceiver/config.go
Comment (Aneurysm9) [reply receiver/prometheusreceiver/config.go:48]: If you won't fix this in this PR then I want to see https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30135 merged first.

**Comment by Aneurysm9** [comment]
> > API users are still users.
> 
> As per my understanding in the changelog we use `api` and `user` to differentiate between API users and end-user configuration/behavior breaking changes. If you disagree with these usages of the 2 words we can change that.

My point was not that the `api` and `user` changelog differentiation is problematic, it was that a breaking change to API users still needs to be clearly motivated and the impact on affected users considered.  We can't simply say "it's just an API breaking change so NBD".

What is the end goal here?  Is this it, cleaning up configuration parsing and that's it, or is this a prelude to further changes?  If so, what are those changes and how may they affect users (both end-users and API consumers)?

> > Has this been communicated to @open-telemetry/operator-ta-maintainers? I know they're at least one group that is going to have an interest in this.
> 
> I don't see a need to do that, since https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-operator%20prometheusreceiver&type=code doesn't show any usage of the public API.

But [they do](https://github.com/open-telemetry/opentelemetry-operator/blob/main/internal/manifests/collector/config_replace.go) use the public API, albeit through YAML mangling and duplicating type definitions.  For that reason I suspect they would be in favor of this change if they were consulted.  If you're planning further changes beyond this refactoring then it's even more important that such stakeholders are consulted.

File: receiver/prometheusreceiver/config.go
Comment (bogdandrutu) [reply receiver/prometheusreceiver/config.go:48]: This bug was here for more than an year, and you are only causing a lot of issues to rebase this PR. I think it is not ok to push for this.

File: receiver/prometheusreceiver/config_test.go
Comment (bogdandrutu) [reply receiver/prometheusreceiver/config_test.go:169]: You are correct, not sure why I changed.

**Comment by bogdandrutu** [comment]
> But [they do](https://github.com/open-telemetry/opentelemetry-operator/blob/main/internal/manifests/collector/config_replace.go) use the public API, albeit through YAML mangling and duplicating type definitions. For that reason I suspect they would be in favor of this change if they were consulted. If you're planning further changes beyond this refactoring then it's even more important that such stakeholders are consulted.

So they don't use our public API, as mentioned previously. I don't see any import of the `prometheusreceiver` package there. This change is 100% safe for the usage that they have which is changing the yaml config, because as stated in the PR description this PR does not change any yaml configuration.

File: receiver/prometheusreceiver/config_test.go
Comment (bogdandrutu) [reply receiver/prometheusreceiver/config_test.go:255]: Done.

**Comment by bogdandrutu** [comment]
> My point was not that the api and user changelog differentiation is problematic, it was that a breaking change to API users still needs to be clearly motivated and the impact on affected users considered. We can't simply say "it's just an API breaking change so NBD".

Does the PR description solve this?

> What is the end goal here? Is this it, cleaning up configuration parsing and that's it, or is this a prelude to further changes? If so, what are those changes and how may they affect users (both end-users and API consumers)?

Do you want more explanation in the PR description about this change, if yes please let me know what is not covered? If there will be changes affecting the end-users and/or API consumers in the future we will document them when the time comes.

**Comment by bogdandrutu** [comment]
@Aneurysm9 friendly ping

File: receiver/prometheusreceiver/config.go
Comment (bogdandrutu) [reply receiver/prometheusreceiver/config.go:48]: PR merged first. Closing this, please let me know what else you want to see to unblock this PR.

File: receiver/prometheusreceiver/config.go
Comment (Aneurysm9) [inline receiver/prometheusreceiver/config.go:47]:
```suggestion
	if (cfg.PrometheusConfig == nil || len(cfg.PrometheusConfig.ScrapeConfigs) == 0) && cfg.TargetAllocator == nil {
```

This still needs to be fixed here.

**Comment by bogdandrutu** [comment]
@Aneurysm9 friendly ping

**Comment by bogdandrutu** [comment]
@Aneurysm9 friendly ping

File: receiver/prometheusreceiver/config.go
Comment (jpkrohling) [reply receiver/prometheusreceiver/config.go:47]: This looks indeed odd still. There are unit tests failing because of this:

```
receivers::prometheus: no Prometheus scrape_configs or target_allocator set
```

https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/7478660342/job/20354137841?pr=29901

File: receiver/prometheusreceiver/factory.go
Comment (jpkrohling) [inline receiver/prometheusreceiver/factory.go:54]: Did you move it here because this logs a warning message?

**Review by jpkrohling — COMMENTED** [review]
Apart from the failing unit tests due to the change in logic in one validation, this looks good.

Out of curiosity: are you doing this change so that you can build a more specialized Prometheus receiver? If so, this might benefit the Pure Storage receivers as well (cc @chrroberts-pure , @dgoscn)

File: receiver/prometheusreceiver/factory.go
Comment (bogdandrutu) [reply receiver/prometheusreceiver/factory.go:54]: It is only used in this file, so I think is better to co-locate the code in this case, easier to read (not having to jump files).

File: receiver/prometheusreceiver/config.go
Comment (bogdandrutu) [reply receiver/prometheusreceiver/config.go]: I think was a glitch somewhere else. Not sure why failed that way.

**Comment by bogdandrutu** [comment]
>Out of curiosity: are you doing this change so that you can build a more specialized Prometheus receiver? If so, this might benefit the Pure Storage receivers as well (cc @chrroberts-pure , @dgoscn)

For the moment focus on cleaning how we do unmarshal/validation and understand patterns.

File: receiver/prometheusreceiver/config.go
Comment (aknuds1) [inline receiver/prometheusreceiver/config.go:75]: [Nit] Sligthly improved language IMO.
```suggestion
// PromConfig is a redeclaration of promconfig.Config because we need custom unmarshaling
// as prometheus "config" uses `yaml` tags.
```

File: receiver/prometheusreceiver/config.go
Comment (aknuds1) [inline receiver/prometheusreceiver/config.go:138]: [Nit]
```suggestion
// PromHTTPSDConfig is a redeclaration of promHTTP.SDConfig because we need custom unmarshaling
// as prometheus "config" uses `yaml` tags.
```

File: receiver/prometheusreceiver/config.go
Comment (aknuds1) [inline receiver/prometheusreceiver/config.go:148]: [Nit]
```suggestion
	cfgMap["url"] = "http://placeholder" // we have to set it as else marshalling will fail
```

File: receiver/prometheusreceiver/config.go
Comment (aknuds1) [inline receiver/prometheusreceiver/config.go:152]: [Nit]
```suggestion
func unmarshalYAMLStruct(in map[string]any, out any) error {
```

**Review by aknuds1 — APPROVED** [review]
LGTM - left some nits.

File: receiver/prometheusreceiver/config.go
Comment (bogdandrutu) [reply receiver/prometheusreceiver/config.go:75]: Done.

File: receiver/prometheusreceiver/config.go
Comment (bogdandrutu) [reply receiver/prometheusreceiver/config.go:138]: Done.

File: receiver/prometheusreceiver/config.go
Comment (bogdandrutu) [reply receiver/prometheusreceiver/config.go:148]: Done.

File: receiver/prometheusreceiver/config.go
Comment (bogdandrutu) [reply receiver/prometheusreceiver/config.go:152]: Done.

**Review by jpkrohling — APPROVED** [review]
LGTM

**Comment by alolita** [comment]
Points noted @Aneurysm9 Thanks!

**Comment by jpkrohling** [comment]
After careful consideration, I am proceeding to merge this PR in my capacity as a @open-telemetry/collector-contrib-maintainer. I would like to outline the reasons for this decision:

- The modifications presented in this PR, particularly in relation to the config placeholder, represent an improvement to our code base.
- The approach taken in this refactoring adheres to best practices, notably in maintaining the integrity of existing tests. Adjustments have been made solely for the incorporation of new types. This gives me a high level of confidence that the changes are unlikely to introduce new bugs, and may help in identifying any previously untested scenarios.
- Although the motivations and subsequent steps for this change were not entirely clear, the nature of the change is such that I would have endorsed it if it were proposed for a component under my code ownership. It is important to maintain consistent standards across our contributors, particularly when they are not strangers to the project/SIG. In this case, the author has met the same level of quality and commitment we expect from all contributors.

**Pull request merged**
